### PR TITLE
add gradient functions for dim_gather op

### PR DESCRIPTION
**Pull request by doombeaker**
给 `dim_gather` （op_type_name 是这个）算子（新接口导出为 `flow.gather`）添加新版本的后向。

### CODE REVIEW

File: oneflow/core/autograd/gradient_funcs/dim_gather.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+};
+
+Maybe<void> DimGather::Init(const OpExpr& op) {
+  printf("yaochi:Init");
+  const auto* fw_op_expr = dynamic_cast<const UserOpExpr*>(&op);
+  CHECK_NOTNULL_OR_RETURN(fw_op_expr);
+  const std::string& op_name = fw_op_expr->op_name();
+  op_trait_ = std::make_shared<user_op::UserOpConfTrait>(op_name, fw_op_expr->proto());
+
+  dim_ = JUST(op_trait_->GetAttr<int32_t>("dim"));
```
Comment (hjchen2) [inline oneflow/core/autograd/gradient_funcs/dim_gather.cpp:51]: 为了支持动态dim，可以把GetAttr统一放到Capture中。同样也可以参考SparseSoftmaxCrossEntropy里面的用法

File: oneflow/core/autograd/gradient_funcs/dim_gather.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  bw_dim_gather_op_ = JUST(op_expr_helper::DimGatherGradOp(dim_, GradientOpName(op_name)));
+  return Maybe<void>::Ok();
+}
+
+Maybe<void> DimGather::Capture(DimGatherInterpState* ctx, const TensorTuple& inputs,
+                               const TensorTuple& outputs, const AttrMap& attrs) const {
+  printf("yaochi:Capture");
+  ctx->requires_grad = inputs.at(1)->requires_grad();
```
Comment (hjchen2) [inline oneflow/core/autograd/gradient_funcs/dim_gather.cpp:59]: 这里应该是`inputs.at(0)->requires_grad()`，inputs的内容应该是`{input, index}`，我们需要拿到的是input的requires_grad信息。

File: oneflow/core/autograd/gradient_funcs/dim_gather.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  if (!ctx->requires_grad) { return Maybe<void>::Ok(); }
+  ctx->SaveTensorForBackward(inputs.at(0));
+  ctx->SaveTensorForBackward(inputs.at(2));
```
[the remark this answers is not in the document]
Comment (doombeaker) [reply oneflow/core/autograd/gradient_funcs/dim_gather.cpp]: 已修正

File: oneflow/core/autograd/gradient_funcs/dim_gather.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  return Maybe<void>::Ok();
+}
+
+Maybe<void> DimGather::Apply(const DimGatherInterpState* ctx, const TensorTuple& out_grads,
+                             TensorTuple* in_grads) const {
+  printf("yaochi:Apply");
+  if (!ctx->requires_grad) { return Maybe<void>::Ok(); }
+  const auto& index = ctx->SavedTensors().at(0);
+  const auto& like = ctx->SavedTensors().at(2);
```
[the remark this answers is not in the document]
Comment (doombeaker) [reply oneflow/core/autograd/gradient_funcs/dim_gather.cpp]: 已修正

File: oneflow/core/framework/op_expr_helper.cpp [hunk 5]
```diff
@@ -537,5 +537,18 @@ Maybe<one::UserOpExpr> SparseSoftmaxCrossEntropyMsGradOp(const int64_t& depth,
       .Build();
 }
 
+Maybe<one::UserOpExpr> DimGatherGradOp(const int32_t dim) {
+  return DimGatherGradOp(dim, UniqueOpName("dim_gather_grad"));
+}
+Maybe<one::UserOpExpr> DimGatherGradOp(const int32_t dim, const std::string& name) {
+  return one::OpBuilder("dim_scatter_add_like", name)
+      .Input("index")
```
[the remark this answers is not in the document]
Comment (doombeaker) [reply oneflow/core/framework/op_expr_helper.cpp]: 已修正

File: oneflow/core/autograd/gradient_funcs/dim_gather.cpp (hunk 2 above)
Comment (doombeaker) [reply oneflow/core/autograd/gradient_funcs/dim_gather.cpp]: 已修正

File: oneflow/core/autograd/gradient_funcs/dim_gather.cpp (hunk 1 above)
Comment (doombeaker) [reply oneflow/core/autograd/gradient_funcs/dim_gather.cpp]: 已经修改

File: oneflow/core/autograd/gradient_funcs/dim_gather.cpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+class DimGather : public OpExprGradFunction<DimGatherInterpState> {
+ public:
+  Maybe<void> Init(const OpExpr& op) override;
+  Maybe<void> Capture(DimGatherInterpState* ctx, const TensorTuple& inputs,
+                      const TensorTuple& outputs, const AttrMap& attrs) const override;
+  Maybe<void> Apply(const DimGatherInterpState* ctx, const TensorTuple& out_grads,
+                    TensorTuple* in_grads) const override;
+
+ private:
+  std::shared_ptr<user_op::UserOpConfTrait> op_trait_;
```
[the remark this answers is not in the document]
Comment (doombeaker) [reply oneflow/core/autograd/gradient_funcs/dim_gather.cpp]: 已经修改

File: oneflow/python/test/modules/test_gather.py [hunk 7]
```diff
@@ -19,7 +19,20 @@
...
+        outcoord = [*outcoord]
+        outcoord[dim] = index[np.unravel_index(srcidx, index.shape)]
+        output_offset = np.ravel_multi_index(outcoord, outshape)
+        output[np.unravel_index(output_offset, outshape)] += src[
+            np.unravel_index(srcidx, src.shape)
+        ]
+    return output
+
+
+def _gather_numpy(input, index, dim):
```
Comment (doombeaker) [inline oneflow/python/test/modules/test_gather.py:35]: 为了一致，并且原有 `gather_numpy` 仅在当前测试文件使用，所以添加了下划线前缀

Comment (doombeaker) [inline oneflow/python/test/modules/test_gather.py:35]: 为了一致，并且原有 `gather_numpy` 仅在当前测试文件使用，所以添加了下划线前缀

File: oneflow/core/autograd/gradient_funcs/dim_gather.cpp [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  CHECK_NOTNULL_OR_RETURN(fw_op_expr);
+  const std::string& op_name = fw_op_expr->op_name();
+  bw_dim_gather_op_ = JUST(op_expr_helper::DimScatterAddLikeOp(0, GradientOpName(op_name)));
+  return Maybe<void>::Ok();
+}
+
+Maybe<void> DimGather::Capture(DimGatherInterpState* ctx, const TensorTuple& inputs,
+                               const TensorTuple& outputs, const AttrMap& attrs) const {
+  ctx->requires_grad = inputs.at(0)->requires_grad();
+  if (!ctx->requires_grad) { return Maybe<void>::Ok(); }
```
Comment (BBuf) [inline oneflow/core/autograd/gradient_funcs/dim_gather.cpp:54]: 这里应该没有必要判断，因为你在Apply里面会也有判断，可以删掉这个

File: oneflow/core/autograd/gradient_funcs/dim_gather.cpp [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+  ctx->SaveTensorForBackward(inputs.at(0));
+
+  ComposedAttrMap composed_attrs(attrs, base_attrs_);
+  ctx->dim = JUST(composed_attrs.GetAttr<int32_t>("dim"));
+  return Maybe<void>::Ok();
+}
+
+Maybe<void> DimGather::Apply(const DimGatherInterpState* ctx, const TensorTuple& out_grads,
+                             TensorTuple* in_grads) const {
+  if (!ctx->requires_grad) { return Maybe<void>::Ok(); }
```
Comment (BBuf) [inline oneflow/core/autograd/gradient_funcs/dim_gather.cpp:66]: 这里补一行`CHECK_EQ_OR_RETURN(out_grads.size(), 1);`吧。

File: oneflow/core/autograd/gradient_funcs/dim_gather.cpp [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+ private:
+  AttrMap base_attrs_;
+  std::shared_ptr<OpExpr> bw_dim_gather_op_;
+};
+
+Maybe<void> DimGather::Init(const OpExpr& op) {
+  const UserOpExpr* fw_op_expr = dynamic_cast<const UserOpExpr*>(&op);
+  CHECK_NOTNULL_OR_RETURN(fw_op_expr);
+  const std::string& op_name = fw_op_expr->op_name();
```
Comment (hjchen2) [inline oneflow/core/autograd/gradient_funcs/dim_gather.cpp:46]: `base_attrs_`需要在这里初始化一下，因为属性还是有可能在build的时候指定的。

File: oneflow/core/autograd/gradient_funcs/dim_gather.cpp (hunk 8 above)
Comment (hjchen2) [reply oneflow/core/autograd/gradient_funcs/dim_gather.cpp:54]: 这里加一下也可以，可以跳过下面的代码

File: oneflow/core/autograd/gradient_funcs/dim_gather.cpp (hunk 9 above)
Comment (doombeaker) [reply oneflow/core/autograd/gradient_funcs/dim_gather.cpp:66]: 已增加

File: oneflow/core/autograd/gradient_funcs/dim_gather.cpp (hunk 10 above)
Comment (doombeaker) [reply oneflow/core/autograd/gradient_funcs/dim_gather.cpp:46]: 已修改

**Review by BBuf — APPROVED** [review]
